x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/Count.java

```diff
@@ -87,4 +93,30 @@ public Nullability nullable() {
     protected TypeResolution resolveType() {
         return EsqlTypeResolutions.isExact(field(), sourceText(), DEFAULT);
     }
+
+    @Override
+    public Expression surrogate() {
+        var s = source();
+        var field = field();
+
+        if (field.foldable()) {
+            if (field instanceof Literal l) {
```

Comment:
Would love to just fold, but folding here is not safe unless we check if this will fold to `null`, first. That would require duplicating logic from `FoldNull`, though, which I'd prefer to avoid.

(Our evaluators cannot deal with null literals in general, and folding here results in the following exception when running e.g. `count(to_double(null))`:
```
illegal data type [null]
org.elasticsearch.xpack.esql.EsqlIllegalArgumentException: illegal data type [null]
	at __randomizedtesting.SeedInfo.seed([DBCF40F13A40EC92:539B7F2B94BC816A]:0)
	at org.elasticsearch.xpack.esql.CsvTests.stats.Asdf(stats.csv-spec:1708)
	at org.elasticsearch.xpack.esql.EsqlIllegalArgumentException.illegalDataType(EsqlIllegalArgumentException.java:43)
	at org.elasticsearch.xpack.esql.EsqlIllegalArgumentException.illegalDataType(EsqlIllegalArgumentException.java:39)
	at org.elasticsearch.xpack.esql.expression.function.scalar.convert.AbstractConvertFunction.evaluator(AbstractConvertFunction.java:65)
	at org.elasticsearch.xpack.esql.expression.function.scalar.convert.AbstractConvertFunction.toEvaluator(AbstractConvertFunction.java:105)
	at org.elasticsearch.xpack.esql.evaluator.mapper.EvaluatorMapper.fold(EvaluatorMapper.java:48)
	at org.elasticsearch.xpack.esql.expression.function.scalar.EsqlScalarFunction.fold(EsqlScalarFunction.java:30)
```
)

Reply:
Consider the rare case of count(1+1) - there might be count(1+null). This can be moved to a separate issue though to not block this one.

Reply:
It makes sense to add more tests to account for these cases where an expression is evaluated, esp. to `null`. I'll do that in another PR, as this one should be backported to 8.14 as soon as possible.

These cases are covered in this PR, it's just that they result in an extra eval+projection.

```
| STATS count(1+null)
```

becomes something equivalent to

```
| STATS x = count(*)
| EVAL `count(1+null)` = COALESCE(MV_COUNT(1+null), 0) * x
| KEEP `count(1+null)`
```

which after folding becomes a multiplication of `count(*)` with 0. If we wanted to, we could further optimize this case to avoid the `count(*)` as well.

Reply:
Opened a PR with additional tests: https://github.com/elastic/elasticsearch/pull/107888